components/generators/argocd/__init__.py

```diff
@@ -0,0 +1,354 @@
+        Allow manifests to deploy from any Git repos
+        """
+        if argocd_project.source_repos:
+            self.root.spec.sourceRepos = argocd_project.source_repos
+
+        """
+        Only permit applications to deploy to the namespace in the same cluster
+        """
+        if argocd_project.destinations:
+            self.root.spec.destinations = argocd_project.destinations
+
+        """
+        Deny all cluster-scoped resources from being created, except for Namespace
+        """
+        if argocd_project.cluster_resource_whitelist:
+            self.root.spec.clusterResourceWhitelist = argocd_project.cluster_resource_whitelist
+
+        """
+        Allow all namespaced-scoped resources to be created, except for ResourceQuota, LimitRange, NetworkPolicy
+        """
+        if argocd_project.namespace_resource_blacklist:
+            self.root.spec.namespaceResourceBlacklist = argocd_project.namespace_resource_blacklist
+
+        """
+        Deny all namespaced-scoped resources from being created, except for Deployment and StatefulSet
+        """
+        if argocd_project.namespace_resource_whitelist:
+            self.root.spec.namespaceResourceWhitelist = argocd_project.namespace_resource_whitelist
+
+        """
+        Enables namespace orphaned resource monitoring.
+        """
+        if argocd_project.orphaned_resources:
+            self.root.spec.orphanedResources = argocd_project.orphaned_resources
+
+        """
+        Roles
+        """
+        if argocd_project.roles:
+            self.root.spec.roles = argocd_project.roles
+
+
+class ArgoCDApplication(k8s.Base):
+    def new(self):
+        self.kwargs.apiVersion = 'argoproj.io/v1alpha1'
+        self.kwargs.kind = 'Application'
+        """
+        Add a this finalizer ONLY if you want these to cascade delete
+        """
+        self.kwargs.finalizers = list('resources-finalizer.argocd.argoproj.io')
+        super().new()
+        self.need('name')
+
+    def body(self):
+        super().body()
+
+        """
+        You'll usually want to add your resources to the argocd namespace.
+        """
+        self.add_namespace(inv.parameters.argocd_namespace)
+
+        argocd_application = self.kwargs.argocd_application
+
+        self.add_annotations(argocd_application.get('annotations', {}))
+        self.add_labels(argocd_application.get('labels', {}))
+
+        """
+        The project the argocd_application belongs to.
+        """
+        self.root.spec.project = argocd_application.project
+        """
+        The destination in which Namespace the application should be deployed
+        """
+        self.root.spec.destination = argocd_application.destination
+        """
+        Source of the application manifests
+        """
+        if argocd_application.source:
+            self.root.spec.source = argocd_application.source
+
+"""
+The following classes are required to generate Secrets + ConfigMaps
+"""
+class SharedConfig():
```

Comment:
Hmm this is copied from the k8s generator, we must find a better way to import it here instead of duplicating. Can we add a TODO comment?